solutions/week1/e03

marcin-czaniecki opened the pull request.

marcin-czaniecki committed:
update .gitignore

marcin-czaniecki committed:
Nodemon and ts-node added to the project

marcin-czaniecki committed:
Files for items and cart

marcin-czaniecki committed:
Item class created

marcin-czaniecki committed:
NowItem class created

marcin-czaniecki committed:
FreeItem class created

marcin-czaniecki committed:
AuctionItem class created

marcin-czaniecki committed:
Cart class created

marcin-czaniecki committed:
fix updateById

marcin-czaniecki committed:
Show cart work, tsconfig added to project

marcin-czaniecki committed:
update tsconfig

michaljabi commented on week-1-ts-ecosystem/w01e03-object-oriented-structure/src/models/Item.ts:
```diff
@@ -0,0 +1,13 @@
+export abstract class Item {
```
Piękny pomysł żeby `Item` było abstrakcyjną bazą modeli danych 👍

michaljabi commented on week-1-ts-ecosystem/w01e03-object-oriented-structure/src/models/Item.ts:
```diff
@@ -0,0 +1,13 @@
+  name: string;
+  amount: number;
+  constructor(name: string, amount: number) {
+    this.id = Math.round(Math.random() * Math.pow(10, 10)).toString(16);
```
polecam użycie biblioteki np. `uuid` - wtedy będzie bardziej `pro` :) 

Dodatkowo zauważ że w tym układzie `id` może być `readonly`

michaljabi commented on week-1-ts-ecosystem/w01e03-object-oriented-structure/src/models/FreeItem.ts:
```diff
@@ -0,0 +1,12 @@
+
+export class FreeItem extends Item {
+  value: undefined;
+  currency: undefined;
```
nie wiem czy tutaj nie dałbym `void` jako podkreślenie że tych wartości nie obsługujemy.

W dalszej perspektywie do zasatnowienia się czy Twoja klasa abstakcyjna `Item` nie robi za dużo.
w sensie czy przechowywanie w niej `value` oraz `currency` ma sens.

Zobacz. Naszą "odmienną" klasą jest jedynie `FreeItem` to może warto do `Item` dodać kolejnego abstrakta jako `ValuedItem` (dziedziczącego po `Item`) i on by miał pola `value` + `currency`

Wtedy dziedziczymy po `ValuedItem` w przypadku `AuctionItem` oraz `NowItem`

michaljabi commented on week-1-ts-ecosystem/w01e03-object-oriented-structure/src/models/NowItem.ts:
```diff
@@ -0,0 +1,16 @@
+import { Item } from "./Item";
+
+export class NowItem extends Item {
```
Zmienił bym nazwę na `BuyNowItem`. ale to szczegół.

michaljabi commented on week-1-ts-ecosystem/w01e03-object-oriented-structure/src/controllers/Cart.ts:
```diff
@@ -0,0 +1,53 @@
+import { Item } from "../models/Item";
+
+export class Cart<Product extends Item> {
```
Tutaj poszedłbym zgodnie z konwencją. Nazwa generyka z dużych:

```typescript
export class Cart<PRODUCT extends Item> { 
//... 
}
```

michaljabi commented on week-1-ts-ecosystem/w01e03-object-oriented-structure/src/controllers/Cart.ts:
```diff
@@ -0,0 +1,53 @@
+      this.producs.set(id, product);
+      return;
+    }
+    console.log("Nie znaleziono produktu o podanym id");
```
Może "mocniej" - zamiast "efektu ubocznego" rzucić wyjątek ?

michaljabi commented on week-1-ts-ecosystem/w01e03-object-oriented-structure/src/controllers/Cart.ts:
```diff
@@ -0,0 +1,53 @@
+import { Item } from "../models/Item";
+
+export class Cart<Product extends Item> {
+  private producs: Map<string, Product> = new Map();
```
fajny pomysł, trzymania wpisów jako mapa

michaljabi commented on week-1-ts-ecosystem/w01e03-object-oriented-structure/src/controllers/Cart.ts:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  getAmount() {
+    return [...this.producs].reduce((acc, [, product]) => acc + product.amount, 0);
```
Czemu tutaj taka komplikacja zamiast `[...this.products.values()]` ?
Albo lepiej żeby sie nie powtarzać:

`this.getAll().reduce .... `

michaljabi commented on week-1-ts-ecosystem/w01e03-object-oriented-structure/src/controllers/Cart.ts:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  getAll() {
+    return [...this.producs.values()];
```
Pamiętaj że da się też bardziej "implicit" pokazać co robisz:

`Array.from(this.products.values())`

michaljabi commented on week-1-ts-ecosystem/w01e03-object-oriented-structure/src/controllers/Cart.ts:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  getFullPrice() {
+    return [...this.producs].reduce((acc, [, product]) => acc + (product?.value || 0) * product.amount, 0).toFixed(2);
```
j.w.

michaljabi commented on week-1-ts-ecosystem/w01e03-object-oriented-structure/src/models/NowItem.ts:
```diff
@@ -0,0 +1,16 @@
+    this.fastShopping = true;
+  }
+  print(): void {
+    console.log(`nazwa: ${this.name}, cena: ${this.value}, ilość: ${this.amount}, wartość: ${this.value * this.amount}`);
```
Ta linia oraz logika zapisu w koszyku:

```
getAmount() {....}
```

sugerują Ci że warto dodać getter do `Item` oznaczający jego wartość jako:
```
get totalValue() {
  return this.value * this.amount
}
```

michaljabi reviewed (changes requested):
Super.

Widać gdzie są modele danych, jakie są, co obsługuje koszyk.
Wszystko na pierwszy rzut oka.

Brak nadmiarowej logiki i powtórzeń w `interface` + `class`

Jedynie drobne moje sugestie do przeanalizowania @janossik 👍

marcin-czaniecki committed:
added uuid

marcin-czaniecki committed:
fixed uuid

marcin-czaniecki committed:
added new class ("ValuedItem")

marcin-czaniecki committed:
changed NowItem on BuyNowItem

marcin-czaniecki committed:
removed ValuedItem and changed Cart

ValuedItem caused problems with the cart for free items, correct implementation would require the implementation of two carts Cart and CartWithValuedItems

marcin-czaniecki commented:
ValuedItem powodował problemy z koszykiem dla darmowych przedmiotów, poprawna implementacja wymagałaby zaimplementowania dwóch koszyków Cart i CartWithValuedItems
